src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs

```diff
@@ -0,0 +1,160 @@
+
+        private static readonly ImmutableArray<(string MethodName, DiagnosticDescriptor Rule)> s_methodMetadataNames = ImmutableArray.Create(
+            ("OfType", OfTypeRule),
+            ("Cast", CastRule)
+        );
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+            => ImmutableArray.Create(OfTypeRule, CastRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemLinqEnumerable, out var enumerableType))
+                {
+                    return;
+                }
+
+                var methodRuleDictionary = s_methodMetadataNames
+                    .SelectMany(m => enumerableType.GetMembers(m.MethodName)
+                                                   .OfType<IMethodSymbol>()
+                                                   .Where(method => method.TypeParameters.HasExactly(1)
+                                                                 && method.Parameters.HasExactly(1))
+                                                   .Select(method => (method, m.Rule)))
+                    .ToImmutableDictionary(key => key.method, v => v.Rule, SymbolEqualityComparer.Default);
+
+                if (methodRuleDictionary.IsEmpty)
+                {
+                    return;
+                }
+
+                context.RegisterOperationAction(context =>
+                {
+                    var invocation = (IInvocationOperation)context.Operation;
+
+                    var targetMethod = (invocation.TargetMethod.ReducedFrom ?? invocation.TargetMethod).OriginalDefinition;
+
+                    if (!methodRuleDictionary.TryGetValue(targetMethod, out var rule))
+                    {
+                        return;
+                    }
+
+                    var instanceArg = invocation.GetInstance(); // "this" argument of an extension method
+
+                    // because the type of the parameter is actually the non-generic IEnumerable,
+                    // we have to reach back through conversion operator(s) to get the "real" type of the argument
+                    if (instanceArg is not IConversionOperation conversionOperation)
+                    {
+                        return;
+                    }
+
+                    var argIEnumerableType = conversionOperation.Operand.Type.AllInterfaces.SingleOrDefault(t => t.OriginalDefinition.SpecialType == SpecialType.System_Collections_Generic_IEnumerable_T);
+                    if (argIEnumerableType is null)
+                    {
+                        return;
+                    }
+
+                    var castFrom = argIEnumerableType.TypeArguments[0];
```

Comment:
Can the type argument count be 0 for invalid code? I'd rather be defensive and check `argIEnumerableType.TypeArguments.IsEmpty` in the above bail out check.